src/gromacs/nblib/interactions.cpp

```diff
@@ -0,0 +1,161 @@
+ * consider that scientific software is very special. Version
+ * control is crucial - bugs must be traceable. We will be happy to
+ * consider code for inclusion in the official distribution, but
+ * derived work must not be called official GROMACS. Details are found
+ * in the README & COPYING files - if they are missing, get the
+ * official version at http://www.gromacs.org.
+ *
+ * To help us fund GROMACS development, we humbly ask that you cite
+ * the research papers on the package. Check out http://www.gromacs.org.
+ */
+/*! \file
+ * \brief
+ * Implements nblib particle-types interactions
+ *
+ * \author Victor Holanda <victor.holanda@cscs.ch>
+ * \author Joe Jordan <ejjordan@kth.se>
+ * \author Prashanth Kanduri <kanduri@cscs.ch>
+ * \author Sebastian Keller <keller@cscs.ch>
+ * \author Artem Zhmurov <zhmurov@gmail.com>
+ */
+
+#include "gmxpre.h"
+
+#include "interactions.h"
+
+#include "gromacs/utility/exceptions.h"
+#include "gromacs/utility/stringutil.h"
+
+namespace nblib
+{
+
+namespace detail
+{
+real combineNonbondedParameters(real v, real w, CombinationRule combinationRule)
+{
+    if (combinationRule == CombinationRule::Geometric)
+    {
+        return sqrt(v * w);
+    }
+    else
+    {
+        throw gmx::InvalidInputError("unknown LJ Combination rule specified\n");
+    }
+}
+
+} // namespace detail
+
+void ParticleTypesInteractions::add(ParticleType particleType, C6 c6, C12 c12)
+{
+    if (singleParticleInteractionsMap_.count(particleType.name()) == 0)
+    {
+        singleParticleInteractionsMap_[particleType.name()] = std::make_tuple(c6, c12);
+        particleTypesSet_.insert(particleType.name());
+    }
+    else
+    {
+        std::string message = gmx::formatString("Attempting to add nonbonded interaction %s twice",
+                                                particleType.name().c_str());
+        GMX_THROW(gmx::InvalidInputError(message));
+    }
+}
+
+void ParticleTypesInteractions::add(ParticleType particleType1, ParticleType particleType2, C6 c6, C12 c12)
+{
+    auto interactionKey         = std::make_tuple(particleType1.name(), particleType2.name());
+    auto possibleInteractionKey = std::make_tuple(particleType2.name(), particleType1.name());
+    if (twoParticlesInteractionsMap_.count(interactionKey) == 0)
+    {
+        std::string message =
+                gmx::formatString("Attempting to add nonbonded interaction between %s %s twice",
+                                  particleType1.name().c_str(), particleType2.name().c_str());
+        GMX_RELEASE_ASSERT(twoParticlesInteractionsMap_.count(possibleInteractionKey) == 0,
+                           message.c_str());
+
+        twoParticlesInteractionsMap_[interactionKey]         = std::make_tuple(c6, c12);
+        twoParticlesInteractionsMap_[possibleInteractionKey] = std::make_tuple(c6, c12);
+
+        particleTypesSet_.insert(particleType1.name());
+        particleTypesSet_.insert(particleType2.name());
+    }
```

Comment:
Similarly to the single-particle case, this should throw if we are trying to overwrite the parameters with different values (but only for this map, we should be able to overwrite the combination rule)

Reply:
done